node/src/components/consensus/era_supervisor/tests.rs

```diff
@@ -169,105 +219,196 @@ fn new_test_chainspec(stakes: Vec<(PublicKey, u64)>) -> Chainspec {
+
+    block
+}
+
+#[tokio::test]
+async fn cross_era_slashing() -> Result<(), Error> {
+    let mut rng = TestRng::new();
+
+    let alice_sk = SecretKey::random(&mut rng);
+    let alice_pk = PublicKey::from(&alice_sk);
+    let bob_sk = SecretKey::random(&mut rng);
+    let bob_pk = PublicKey::from(&bob_sk);
+
+    let (mut es, effects) = {
+        let chainspec = new_test_chainspec(vec![(alice_pk, 10), (bob_pk, 100)]);
+        let config = Config {
+            secret_key_path: External::Loaded(alice_sk),
+        };
+
+        let registry = Registry::new();
+        let reactor = MockReactor::new();
+        let effect_builder = EffectBuilder::new(EventQueueHandle::new(reactor.scheduler));
+
+        // Initialize the era supervisor. There are two validators, Alice and Bob. This instance,
+        // however, is only a passive observer.
+        EraSupervisor::new(
+            Timestamp::now(),
+            WithDir::new("tmp", config),
+            effect_builder,
+            vec![
+                (alice_pk, Motes::new(10.into())),
+                (bob_pk, Motes::new(100.into())),
+            ],
+            &chainspec,
+            Default::default(), // genesis state root hash
+            &registry,
+            Box::new(MockProto::new_boxed),
+            &mut rng,
+        )?
+    };
+    assert!(effects.is_empty());
+
+    let parent_hash = BlockHash::from(Digest::from([0; Digest::LENGTH]));
+    let block = propose_and_finalize(&mut es, bob_pk, vec![alice_pk], parent_hash, &mut rng).await;
+    let fb: FinalizedBlock = block.header().clone().into();
     let expected_fb = FinalizedBlock::new(
-        proto_block,
-        timestamp,
+        fb.proto_block().clone(),
+        fb.timestamp(),
         None, // not the era's last block
         EraId(0),
         0, // height
         bob_pk,
     );
-    assert_eq!(expected_fb, reactor.expect_execute().await);
-    assert_eq!(expected_fb, reactor.expect_finalized().await);
+    assert_eq!(expected_fb, fb);
+    let parent_hash = *block.hash();
+
+    let block = propose_and_finalize(&mut es, bob_pk, vec![], parent_hash, &mut rng).await;
+    let fb: FinalizedBlock = block.header().clone().into();
+    let expected_fb = FinalizedBlock::new(
+        fb.proto_block().clone(),
+        fb.timestamp(),
+        Some(EraEnd {
+            equivocators: vec![alice_pk],
+            rewards: Default::default(),
+        }), // the era's last block
+        EraId(0),
+        1, // height
+        bob_pk,
+    );
+    assert_eq!(expected_fb, fb);
+    let parent_hash = *block.hash();
+
+    let block = propose_and_finalize(&mut es, bob_pk, vec![], parent_hash, &mut rng).await;
+    let fb: FinalizedBlock = block.header().clone().into();
+    let expected_fb = FinalizedBlock::new(
+        fb.proto_block().clone(),
+        fb.timestamp(),
+        None, // not the era's last block
+        EraId(1),
+        2, // height
+        bob_pk,
+    );
+    assert_eq!(expected_fb, fb);
```

Comment:
This test is callsed `cross_era_slashing` but I don't see any of it happening :thinking:

Reply:
> This still doesn't fully test cross-era slashing, but that should be doable in the next (hopefully smaller) PR.

I see now. So what should I review exactly?

Reply:
Good point, I'll rename it. So far it only proposes and finalizes three blocks, and requests evidence and switches eras.